tests/unit-tests/input/evdev/test_libinput_device.cpp

```diff
@@ -553,6 +553,110 @@ TEST_F(LibInputDeviceOnMouse, process_event_handles_scroll)
+                                        pressure, orientation);
+    env.mock_libinput.setup_touch_up_event(fake_device, event_time_1, slot2);
+
+    //Add another valid touch event to make sure the frame is really not processed
+    env.mock_libinput.setup_touch_event(fake_device, LIBINPUT_EVENT_TOUCH_DOWN, event_time_1, slot3, 1, 1, major, minor,
+                                        pressure, orientation);
+    env.mock_libinput.setup_touch_up_event(fake_device, event_time_1, slot3);
+
+    env.mock_libinput.setup_touch_frame(fake_device, event_time_1);
+    process_events(touch_screen);
+}
+
+TEST_F(LibInputDeviceOnTouchScreen, process_event_does_not_ignore_valid_touch_events_on_coordinate_0_0)
+{
+    MirTouchId slot1 = 0, slot2 = 1;
+    float major = 0;
+    float minor = 0;
+    float pressure = 0.0f;
+    float x = 0;
+    float y = 0;
+    float orientation = 0;
+
+    InSequence seq;
+    EXPECT_CALL(mock_builder, touch_event(_, _)).Times(1);
+    EXPECT_CALL(mock_sink, handle_input(_)).Times(1);
+
+    touch_screen.start(&mock_sink, &mock_builder);
+    env.mock_libinput.setup_touch_event(fake_device, LIBINPUT_EVENT_TOUCH_DOWN, event_time_1, slot1, x, y, major, minor,
+                                        pressure, orientation);
+    env.mock_libinput.setup_touch_up_event(fake_device, event_time_1, slot1);
+    env.mock_libinput.setup_touch_event(fake_device, LIBINPUT_EVENT_TOUCH_DOWN, event_time_1, slot2, 1, 1, major, minor,
+                                        pressure, orientation);
+    env.mock_libinput.setup_touch_up_event(fake_device, event_time_1, slot2);
+    env.mock_libinput.setup_touch_frame(fake_device, event_time_1);
+    process_events(touch_screen);
+}
+
+TEST_F(LibInputDeviceOnTouchScreen, process_event_with_two_downs_but_one_up)
+{
+    MirTouchId slot1 = 0;
+    float major = 0;
+    float minor = 0;
+    float pressure = 0.0f;
+    float x = 1;
+    float y = 1;
+    float orientation = 0;
+
+    touch_screen.start(&mock_sink, &mock_builder);
+
+    // Detect single touch down then single up
+    EXPECT_CALL(mock_builder, touch_event(_, _)).Times(2);
+    EXPECT_CALL(mock_sink, handle_input(_)).Times(2);
+
+    // Send first down
+    env.mock_libinput.setup_touch_event(fake_device, LIBINPUT_EVENT_TOUCH_DOWN, event_time_1, slot1, x, y, major, minor,
+                                        pressure, orientation);
+    env.mock_libinput.setup_touch_frame(fake_device, event_time_1);
+
+    // Send second down
+    env.mock_libinput.setup_touch_event(fake_device, LIBINPUT_EVENT_TOUCH_DOWN, event_time_2, slot1, x, y, major, minor,
+                                        pressure, orientation);
+    env.mock_libinput.setup_touch_frame(fake_device, event_time_2);
+
+    // Send single up
+    env.mock_libinput.setup_touch_up_event(fake_device, event_time_3, slot1);
+    env.mock_libinput.setup_touch_frame(fake_device, event_time_3);
+    process_events(touch_screen);
+}
+
```

Comment:
This test is wrong. We should probably treat the second down as a move and then output the up.

Instead, the code that was committed with it actually generates two down events and then resets the internal state so that the up is ignored.